File: benchmarks/15-optimization/optimization.m
(* Experiment 15 -- Numerical optimisation.
   MEASURES src/findmin.c and src/fit.c: unconstrained minimisation and linear
   least squares.  Checks are the objective VALUE at the minimum, rounded --
   invariant under which descent path was taken. *)

Get["../harness.m"];
Get["../data.m"];

require[{"FindMinimum", "FindMaximum", "Minimize", "NMinimize", "Fit",
         "FindFit", "LeastSquares"}];

(* 1. A quadratic bowl: the easiest possible case, and the floor. *)
bench["FindMinimum quadratic", FindMinimum[(x - 3)^2 + 2, {x, 0}];];
check["FindMinimum quadratic",
  Round[10^6 First[FindMinimum[(x - 3)^2 + 2, {x, 0}]]]];

(* 2. Rosenbrock in 2-D: the classic curved-valley test. *)
bench["FindMinimum Rosenbrock 2-D",
  FindMinimum[(1 - x)^2 + 100 (y - x^2)^2, {{x, -1}, {y, 1}}];];
check["FindMinimum Rosenbrock 2-D",
  Round[10^4 First[FindMinimum[(1 - x)^2 + 100 (y - x^2)^2,
                               {{x, -1}, {y, 1}}]]]];

(* 3. A 1-D transcendental objective. *)
bench["FindMinimum x Sin[x] on [2,6]",
  FindMinimum[x Sin[x], {x, 4}];];
check["FindMinimum x Sin[x] on [2,6]",
  Round[10^4 First[FindMinimum[x Sin[x], {x, 4}]]]];

(* 4. Linear least squares through Fit: a normal-equations solve. *)
(* MACHINE FLOATS, NOT EXACT RATIONALS.  `i/100` in the Wolfram language is an
   exact Rational, so `Fit` over such points does EXACT rational linear algebra
   while the numpy column does float64 -- two different problems, and the first
   run of this experiment reported 239000x because of it, not because of any
   defect.  N[] pins both columns to machine doubles. *)
pts = N[Table[{i, 2 i + 1}, {i, 1, 200}]];
bench["Fit linear, 200 points", Fit[pts, {1, t}, t];];
check["Fit linear, 200 points", Round[10^6 (Fit[pts, {1, t}, t] /. t -> 10)]];

(* 5. Polynomial least squares, degree 5 over 500 points. *)
pts2 = N[Table[{i/100, (i/100)^3 - 2 (i/100) + 1}, {i, 1, 500}]];
bench["Fit degree-5 polynomial, 500 points",
  Fit[pts2, {1, t, t^2, t^3, t^4, t^5}, t];];
check["Fit degree-5 polynomial, 500 points",
  Round[10^4 (Fit[pts2, {1, t, t^2, t^3, t^4, t^5}, t] /. t -> 2)]];

File: benchmarks/15-optimization/optimization.py
#!/usr/bin/env python3
"""Experiment 15 -- Numerical optimisation (scipy.optimize column).

Same five kernels as ``optimization.m``, same order.  Checks are the objective
value at the minimum, rounded -- invariant under which descent path was taken.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import math

import numpy as np
from scipy.optimize import minimize

from harness import bench, check, require

require(["scipy.optimize:minimize", "numpy.polynomial:polynomial",
         "numpy:polyfit"])

TOL = dict(tol=1e-12)


def r(x, p):
    return int(math.floor(x * 10 ** p + 0.5))


q = lambda v: (v[0] - 3.0) ** 2 + 2.0
bench("FindMinimum quadratic", lambda: minimize(q, [0.0], **TOL))
check("FindMinimum quadratic", r(minimize(q, [0.0], **TOL).fun, 6))

ros = lambda v: (1.0 - v[0]) ** 2 + 100.0 * (v[1] - v[0] ** 2) ** 2
bench("FindMinimum Rosenbrock 2-D", lambda: minimize(ros, [-1.0, 1.0], **TOL))
check("FindMinimum Rosenbrock 2-D", r(minimize(ros, [-1.0, 1.0], **TOL).fun, 4))

xs = lambda v: v[0] * math.sin(v[0])
bench("FindMinimum x Sin[x] on [2,6]", lambda: minimize(xs, [4.0], **TOL))
check("FindMinimum x Sin[x] on [2,6]", r(minimize(xs, [4.0], **TOL).fun, 4))

ti = np.arange(1.0, 201.0)
yi = 2.0 * ti + 1.0
bench("Fit linear, 200 points", lambda: np.polyfit(ti, yi, 1))
_c1 = np.polyfit(ti, yi, 1)
check("Fit linear, 200 points", r(float(np.polyval(_c1, 10.0)), 6))

t2 = np.arange(1, 501) / 100.0
y2 = t2 ** 3 - 2.0 * t2 + 1.0
bench("Fit degree-5 polynomial, 500 points", lambda: np.polyfit(t2, y2, 5))
_c2 = np.polyfit(t2, y2, 5)
check("Fit degree-5 polynomial, 500 points", r(float(np.polyval(_c2, 2.0)), 4))
